### ISSUE TITLE
regression 1.49: macro_rules unexpected tokens

### ISSUE THREAD

**Issue by Mark-Simulacrum** [rust-lang/rust#79908]
Unclear on the extent to which these are the same problem:

* https://crater-reports.s3.amazonaws.com/beta-1.49-1/beta-2020-11-26/gh/ErickHdez96.tiger/log.txt
* https://crater-reports.s3.amazonaws.com/beta-1.49-1/beta-2020-11-26/reg/fourier-0.1.0/log.txt

cc @petrochenkov @Aaron1011

**Comment by Aaron1011** [comment]
The `fourier` error looks like an old bug that was fixed by https://github.com/calebzulawski/fourier/pull/17

**Comment by Mark-Simulacrum** [comment]
https://crates.io/crates/fourier only lists 0.1.0 so we should probably nag @calebzulawski about a point release one more time :)

**Comment by calebzulawski** [comment]
Whoops, looks like I half-fixed it.  One of my dependencies is too strict.  It does look like the same error so probably not a regression...

**Comment by Mark-Simulacrum** [comment]
cc @petrochenkov

**Comment by camelid** [comment]
@apiraino and I both feel unclear about this: Is this a Rust regression or a regression in a library that was then picked up by Crater?

**Comment by Mark-Simulacrum** [comment]
This is a regression in the Rust compiler.

**Comment by camelid** [comment]
The `tiger` error seems to me like fallout from implementing `inline_const`. I'm guessing the issue is that it's expecting a block but immediately gets a literal. Not sure how this is supposed to behave though.

-----

Reposting the error here:

```
[INFO] [stdout] error: expected expression, found keyword `const`
[INFO] [stdout]    --> src/ir.rs:481:23
[INFO] [stdout]     |
[INFO] [stdout] 347 |     (exp $e:expr) => {
[INFO] [stdout]     |          ------- while parsing argument for this `expr` macro fragment
[INFO] [stdout] ...
[INFO] [stdout] 481 |             stmt!(exp const 1);
[INFO] [stdout]     |                       ^^^^^ expected expression
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: expected expression, found keyword `const`
[INFO] [stdout]    --> src/ir.rs:482:23
[INFO] [stdout]     |
[INFO] [stdout] 347 |     (exp $e:expr) => {
[INFO] [stdout]     |          ------- while parsing argument for this `expr` macro fragment
[INFO] [stdout] ...
[INFO] [stdout] 482 |             stmt!(exp const 2);
[INFO] [stdout]     |                       ^^^^^ expected expression
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: expected expression, found keyword `const`
[INFO] [stdout]    --> src/ir.rs:483:23
[INFO] [stdout]     |
[INFO] [stdout] 347 |     (exp $e:expr) => {
[INFO] [stdout]     |          ------- while parsing argument for this `expr` macro fragment
[INFO] [stdout] ...
[INFO] [stdout] 483 |             stmt!(exp const 3);
[INFO] [stdout]     |                       ^^^^^ expected expression
[INFO] [stdout] 
[INFO] [stdout] 
[INFO] [stdout] error: aborting due to 3 previous errors
```

**Comment by camelid** [comment]
But I'm not quite sure; the span seems potentially wrong.

**Comment by spastorino** [comment]
@rustbot ping cleanup

would be nice to get an MCVE

**Comment by camelid** [comment]
Somewhat minimized ([playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=d01b6ec160d95f1dad061f8f8d079b94)):

```rust
macro_rules! exp {
    (const $n:expr) => {
        Box::new(Exp::Const($n))
    };
}

macro_rules! stmt {
    (exp $e:expr) => {
        Box::new(Stmt::Exp($e))
    };
    (exp $($t:tt)+) => {
        Box::new(Stmt::Exp(exp!($($t)+)))
    };
}

macro_rules! seq {
    ($stmt1:expr; $stmt2:expr $(;)?) => {
        Box::new(Stmt::Seq($stmt1, $stmt2))
    };
    ($stmt1:expr; $stmt2:expr; $($stmt:expr);+ $(;)?) => {
        seq!(seq!($stmt1; $stmt2); $($stmt);+)
    };
}

fn main() {
    let seq = seq!(
        stmt!(exp const 1);
        stmt!(exp const 2);
        stmt!(exp const 3);
    );
}
```

On **stable**, it produces a bunch of semantic errors (which make sense since I removed a lot of type definitions), but on **beta** it produces:

```
error: expected expression, found keyword `const`
  --> src/main.rs:99:19
   |
47 |     (exp $e:expr) => {
   |          ------- while parsing argument for this `expr` macro fragment
...
99 |         stmt!(exp const 1);
   |                   ^^^^^ expected expression

error: expected expression, found keyword `const`
   --> src/main.rs:100:19
    |
47  |     (exp $e:expr) => {
    |          ------- while parsing argument for this `expr` macro fragment
...
100 |         stmt!(exp const 2);
    |                   ^^^^^ expected expression

error: expected expression, found keyword `const`
   --> src/main.rs:101:19
    |
47  |     (exp $e:expr) => {
    |          ------- while parsing argument for this `expr` macro fragment
...
101 |         stmt!(exp const 3);
    |                   ^^^^^ expected expression
```

**Comment by spastorino** [comment]
Assigning `P-high` as [discussed as part of the Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/I-prioritize.20.2379908.20regression.201.2E49.3A.20macro_rules.20unexpected.E2.80.A6) and removing `I-prioritize`.

**Comment by camelid** [comment]
Much more minimized:

```rust
macro_rules! exp {
    (const $n:expr) => {
        $n
    };
}

macro_rules! stmt {
    (exp $e:expr) => {
        $e
    };
    (exp $($t:tt)+) => {
        exp!($($t)+)
    };
}

fn main() {
    stmt!(exp const 1);
}
```

([Playground](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=d867b513cae56b3dd6eabb4a25e95009))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error: expected expression, found keyword `const`
  --> src/main.rs:17:15
   |
8  |     (exp $e:expr) => {
   |          ------- while parsing argument for this `expr` macro fragment
...
17 |     stmt!(exp const 1);
   |               ^^^^^ expected expression

warning: unused macro definition
 --> src/main.rs:1:1
  |
1 | / macro_rules! exp {
2 | |     (const $n:expr) => {
3 | |         $n
4 | |     };
5 | | }
  | |_^
  |
  = note: `#[warn(unused_macros)]` on by default

error: aborting due to previous error; 1 warning emitted

error: could not compile `playground`

To learn more, run the command again with --verbose.
```

**Comment by camelid** [comment]
Here's the code again:

```rust
macro_rules! exp {
    (const $n:expr) => {
        $n
    };
}

macro_rules! stmt {
    (exp $e:expr) => {
        $e
    };
    (exp $($t:tt)+) => {
        exp!($($t)+)
    };
}

fn main() {
    stmt!(exp const 1);
}
```

I think the issue is that **on beta** the compiler thinks the `const` in the macro invocation is starting an expression because of the new `inline_const` feature – thus it thinks it's using the first rule in the macro. But then when doing macro expansion, it realizes that it's actually not a valid expression because it's missing braces.

**On stable** this is not an issue because the compiler never thinks that `const` is starting an expression.

I think the fix might be to not treat `const` as the start of an expression if it is not immediately followed by braces.

*See the original `inline_const` PR: #77124.*

cc @spastorino who has been implementing `inline_const`

**Comment by camelid** [comment]
My bisection is in line with my diagnosis:

searched nightlies: from nightly-2020-10-15 to nightly-2020-10-19
regressed nightly: nightly-2020-10-18
searched commits: from https://github.com/rust-lang/rust/commit/e3051d8c24467b54c81da9d9ad70a3e82e106ec1 to https://github.com/rust-lang/rust/commit/043eca7f0b34d12e61c44206beca740628647080
regressed commit: https://github.com/rust-lang/rust/commit/6af9846fcc8797bf97e9fb387385208c2219f3d0

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.0</summary>


Host triple: x86_64-apple-darwin
Reproduce with:
```bash
cargo bisect-rustc --preserve --regress=error --start=2020-10-15 --end=2020-10-19 
```
</details>

**Comment by camelid** [comment]
I'm going to try working on this. If I get stuck I'll release my assignment since this is P-high!

**Comment by camelid** [comment]
So I'm trying to figure out which code is the culprit here. The `inline_const` PR is the one that introduced the regression, but looking at its changes I don't know if it's actual the underlying cause. I think the issue might be in the macro expansion code, but I'm not familiar with it at all.

Since this issue is P-critical and potentially a release blocker, I'm going to release my assignment so that someone more experienced can take it over. Though I'm happy to help if there's some way to do that!

**Comment by camelid** [comment]
Actually, I just noticed that petrochenkov responded to my message asking for help finding the code 😄

**Comment by camelid** [comment]
This should be fixed on the next nightly.

**Comment by Mark-Simulacrum** [comment]
Reopening to track beta fix.

**Comment by Mark-Simulacrum** [comment]
Should be backported to beta as part of https://github.com/rust-lang/rust/pull/80417.

**Comment by camelid** [comment]
I think this can be closed as #80417 has been merged?
